### PR TITLE
Stop full object exposure on the front end

### PR DESCRIPTION
**Pull request by eternoendless**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | 1.7.3.x
| Description?  | This PR introduces object filtering through property whitelisting. This filtering is now applied to:<br>1. Objects being sent to the front office through the "prestashop" object,<br>2. Objects being sent after adding a product to the cart,<br>3. Objects being sent in response to search.<br><br>It also introduces two hooks:<br>1. actionBuildFrontEndObject - Allows to add elements to the "prestashop" object included in javascript<br>2. actionFrontControllerAfterInit - Launched after the "init" method on all front controllers
| Type?         | bug fix
| Category?     | CO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | http://forge.prestashop.com/browse/BOOM-4509
| How to test?  | See ticket

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/8803)

### CODE REVIEW

File: classes/controller/ProductListingFrontController.php [hunk 1]
```diff
@@ -497,24 +500,48 @@ protected function getAjaxProductSearchVariables()
...
-        );
-        foreach ($products as $product_key => $product) {
-            foreach ($product as $product_property => $data) {
-                if (!in_array($product_property, $allowed_properties)) {
-                    unset($products[$product_key][$product_property]);
-                }
-            }
-        }
-        return $products;
+        $filter = (new CollectionFilter())
```
Comment (mickaelandrieu) [inline classes/controller/ProductListingFrontController.php]: you have access to service container in controllers, take a look at config/services/services/front folder :)

File: src/Core/Filter/FilterInterface.php [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+ * @copyright 2007-2018 PrestaShop SA
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+namespace PrestaShop\PrestaShop\Core\Filter;
+
+interface FilterInterface
+{
+    public function filter($subject);
```
Comment (mickaelandrieu) [inline src/Core/Filter/FilterInterface.php]: phpdoc please so you can declare {@inheritdoc} everywhere: wdyt? :)

File: src/Core/Filter/FrontEndObject/ConfigurationFilter.php [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+use PrestaShop\PrestaShop\Core\Filter\HashMapWhitelistFilter;
+
+/**
+ * Filters Configuration objects that will be sent to the client
+ */
+class ConfigurationFilter extends HashMapWhitelistFilter
+{
+    public function __construct()
+    {
+        $whitelist = array(
```
Comment (mickaelandrieu) [inline src/Core/Filter/FrontEndObject/ConfigurationFilter.php:39]: hum... looks like a configurable object... maybe a factory instead of multiples classes?

**Comment by PrestaEdit** [comment]
As I understand, only the properties set inside the white list array will be exposed ? And this list can't be modified by a third party module, isn't it ?

If yes, I think it will be good to rethink this to a black list (and not white one) or allow the third party modules to alter the white list array.

**Comment by mickaelandrieu** [comment]
@PrestaEdit if every filter is called using the service container you'll be able to override them by overriding the service definitions in the module (feature available in 1.7.3+ only). So implementing the right interface will give you full powers and responsabilities.

**Comment by kpodemski** [comment]
@mickaelandrieu what about multiple modules? does one respect another modifications?

**Comment by PrestaEdit** [comment]
You want no more override and you mention... overriding feature. It's kind of no logic, there, no ? :)

File: classes/controller/ProductListingFrontController.php [hunk 4]
```diff
@@ -497,24 +500,48 @@ protected function getAjaxProductSearchVariables()
...
+            ->queue([
+                (new HashMapWhitelistFilter())
```
Comment (mickaelandrieu) [inline classes/controller/ProductListingFrontController.php]: Shouldnt be ProductFilter for consistency?

**Comment by kpodemski** [comment]
I must agree with @PrestaEdit - we need to extend, not override.

File: classes/controller/ProductListingFrontController.php (hunk 4 above)
Comment (mickaelandrieu) [reply classes/controller/ProductListingFrontController.php]: If not, how about create a specific and overiddable class/object/service instead? I feel we should not write configuration in controller :(

**Comment by Quetzacoalt91** [comment]
When I have a global look on this PR, I wonder why these lists are not stored near, or with the `$definition` properties of our object Model, this could avoid for instance several definition of the Product Filter list (i.e in `ProductListingFrontController.php` and `ProductFilter.php`.

Furthermore, we could chose between a whitelist & a blacklist, your product list would be shorter. :)

**Comment by mickaelandrieu** [comment]
@PrestaEdit I don't want people to override a 800 lines controller to add a key in an array called "$filters" ^^

@kpodemski with or without 1.7, when 2 modules override the same things, there is a conflict: this is not something we can change. But with Symfony and services, you could have a module A that override the filter A and a module B that override the filter B. 

In 1.6 "dev like", the module A override the complete controller and if you want to alter anything in it, you can't.

I'm not in favor of extension here, because theses filters shouldnt be used by modules (for me this should be final classes). Why? Because we need to limit the number of classes from Core you use directly in your modules, and instead make you implement the right interfaces: better modules and easier maintenance for both and you :-)

It's my personal opinion ;)

**Comment by PrestaEdit** [comment]
Even for one or two lines, overriding is overrinding. There is not others words about that.

**Comment by mickaelandrieu** [comment]
@Quetzacoalt91 too much responsabilities for object model already if you ask me :D it's a domain model and an active record orm and a table gateway mapper and an Api-like mapper xD

**Comment by mickaelandrieu** [comment]
@PrestaEdit yes, you're right but I never said I'm against of the override of Symfony services available in PrestaShop: this is the main reason we provide them ;)

But you know what happens in 1.6: people override a class AND extends the original one at the same time. And when we make updates of Core, they cry a lot: this is what " i " want to avoid. Providing overridable services definitions that rely on final classes and implement interfaces force module developers to alter only what they need, and to not depend on Core.

This way we limit conflicts between modules (not all, but a lot more than before, you'll agree on that) and you don't need to "guess" from witch "override" folder the class you try to debug came from (painful work if you ask me)

**Comment by PrestaEdit** [comment]
Bad people do that, yes. :)

You don't need to adapt to them, they need to adapt to what they need to do !

Somes modules use an object properties inside somes hooks to alter some others properties, and somes modules add new properties. 

It's why I think the better thing is to allow a module to alter this array, as it.

**Comment by mickaelandrieu** [comment]
@PrestaEdit I'm not against new hooks here, in addition to services: there is nothing wrong with the old good "hooks-way" to do :)

In the end, it's @eternoendless decision :D

**Comment by PrestaEdit** [comment]
So, it's good. It's just to have maybe new hooks there, that I have take the voice here ! :D

**Comment by mickaelandrieu** [comment]
@PrestaEdit I'm getting old, I need time to understand it ahahaha

File: classes/controller/ProductListingFrontController.php (hunk 1 above)
Comment (eternoendless) [reply classes/controller/ProductListingFrontController.php]: Good to know! But how would modules override that?

File: src/Core/Filter/FilterInterface.php (hunk 2 above)
Comment (eternoendless) [reply src/Core/Filter/FilterInterface.php]: I'll add the phpdoc but I don't think I will inheritdoc everywhere since each concrete class has its own parameter and return type.

File: src/Core/Filter/FrontEndObject/ConfigurationFilter.php (hunk 3 above)
Comment (eternoendless) [reply src/Core/Filter/FrontEndObject/ConfigurationFilter.php:39]: > maybe a factory

Like this?

```php
class ConfigurationFilterFactory()
{
    public function build()
    {
        return new HashMapWhitelistFilter(array(...));
    }
}
```

> instead of multiples classes?

🤔 what do you mean? A single factory with multiple factory methods, one for each type?

File: classes/controller/ProductListingFrontController.php (hunk 1 above)
Comment (mickaelandrieu) [reply classes/controller/ProductListingFrontController.php]: You have information in docs you asked me to do :) 

tldr: write a yml file and a class which implement the same interface in the module

File: src/Core/Filter/FrontEndObject/ConfigurationFilter.php (hunk 3 above)
Comment (mickaelandrieu) [reply src/Core/Filter/FrontEndObject/ConfigurationFilter.php:39]: maybe like this

```php
$factory->build($name, $whitelist) // return instance of filter interface
```
I want to extract the configuration from objects... if we do that we can see that we dont really need them, isn't it?

File: src/Core/Filter/FilterInterface.php (hunk 2 above)
Comment (mickaelandrieu) [reply src/Core/Filter/FilterInterface.php]: fair enough :+1:

File: classes/controller/ProductListingFrontController.php (hunk 1 above)
Comment (mickaelandrieu) [reply classes/controller/ProductListingFrontController.php]: a real exemple here: https://github.com/PrestaShop/prestafony-project/blob/master/modules/ps_test/config/services.yml

File: src/Core/Filter/FrontEndObject/ConfigurationFilter.php (hunk 3 above)
Comment (eternoendless) [reply src/Core/Filter/FrontEndObject/ConfigurationFilter.php:39]: > I want to extract the configuration from objects

Oh that's already the case:
```php
(new HashMapWhitelistFilter())->whitelist($whitelist)
```

My original idea was that you can mix & match generic filters in order to filter more complex structures in a more-or-less semantic way. What value would a factory add here?

Specific classes like `ProductFilter` are simply an extended filter class with a hardcoded whitelist. They serve a dual purpose:

1. Provide better semantics for mixed & nested filters
2. Encourage reusability and extensibility

I'll explain more in a bit.

File: classes/controller/ProductListingFrontController.php (hunk 1 above)
Comment (eternoendless) [reply classes/controller/ProductListingFrontController.php]: Ohh, that could be the answer to the extensibility issue I'm struggling with!

Comment (mickaelandrieu) [reply classes/controller/ProductListingFrontController.php]: <3 <3 Always happy to help <3<3

File: src/Core/Filter/FrontEndObject/ConfigurationFilter.php (hunk 3 above)
Comment (mickaelandrieu) [reply src/Core/Filter/FrontEndObject/ConfigurationFilter.php:39]: you're right, I'm frustrated to not have a computer here to play a little bit with your pull request :)

**Comment by mickaelandrieu** [comment]
Thinking about how we can provide Hook(s) , what do you suggest @PrestaEdit?

File: src/Core/Filter/FrontEndObject/CustomerFilter.php [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+                )),
+            "is_logged",
+            "lastname",
+            "newsletter",
+            "newsletter_date_add",
+            "optin",
+            "siret",
+            "website",
+        );
+
```
Comment (kpodemski) [inline src/Core/Filter/FrontEndObject/CustomerFilter.php:59]: @mickaelandrieu 

what about:
`$whitelist = \Hook::exec('actionCustomerFieldsWhiteListModifier', ['whitelist' => &$whitelist]);`

?

**Comment by eternoendless** [comment]
Before anything, thank you all for pitching in! That means that you care about what we are doing 🙂  

## About whitelist vs blacklist

> [@PrestaEdit]: I think it will be good to rethink this to a black list (and not white one) or allow the third party modules to alter the white list array.

> [@Quetzacoalt91]: Furthermore, we could chose between a whitelist & a blacklist, your product list would be shorter. :)

Let's get this said:

**Blacklisting Is A Bad Idea**.

Why? Because it's ***unsafe by default***. You don't know which properties will be added in the future, and you absolutely don't want to obliviously let them leak to the outside world.

Incomplete blacklisting is the very reason I had to make this PR: some people added properties with sensitive data to the Product model on their shop, and found them leaking to the front office!

Blacklisting relies on two principles that are incompatible with PrestaShop:

1. **It needs a stable model**. You can only blacklist if you know what needs to be filtered out. An extensible system means that, by definition, you _cannot_ know what you'll have in it in advance.

2. **It requires Full knowledge** of what has to be done in order to have it working properly (aka the "Don't forget that..." syndrome). If you need to remember, you'll _probably_ forget. Forgot to add your data to the blacklist? You won't realize until you data is already exposed and the damage is done. It's what _actually_ happened.

**Why is whitelisting better?**

- **It's safe by default**. You have to _explicitly_ add it to the list to expose it.
- **It won't bite you in the back if you forget about it**. Your data won't be available if you forget to whitelist it. If you _actually_ need it, you'll notice right away.

Of course, it has the disadvantage of being "a little bit more work" because the list is longer. But I think it's worth it, by a long stretch.

## About override vs extending

> [@kpodemski]: I must agree with @PrestaEdit - we need to extend, not override.

I agree with the need to extend. Override _may_ be useful in certain contexts, so we need to keep that possibility open, but extensibility should always be preferred.

I had already taken the overridable part into account (isolating the filtering part on the controller so it would be easily overridable), as well as some degree of extensibility (generic filters, `whitelist()` allowing to add or replace elements on the whitelist).

I think I can improve on the extensibility part by declaring filters as services, therefore allowing modules to alter them using the service container. For example if you needed to whitelist another field on your Product, instead of overriding the whole ProductFilter, you'll do it like this:

```php
$this->get('product_front_end_filter')->whitelist(array('myCustomProperty'));
```

... with the advantage of all the modules being able to add items to the whitelist without overriding each other!

## About the division of responsibilities

> [@Quetzacoalt91]: I wonder why these lists are not stored near, or with the `$definition` properties of our object Model

For the same reason that I made this separate from the Presenters. The way I see it, there are three different levels of responsibility:

1. **Models** - Source data, persistance
2. **Presenters** - Enriched data, to be used by the back end (eg. templates)
3. **Filters** - Whitelisted data to be exposed to the front end

That's why they have to be kept separate, even if it means duplicating things.

File: classes/controller/ProductListingFrontController.php (hunk 4 above)
Comment (eternoendless) [reply classes/controller/ProductListingFrontController.php]: Yeah you're right, I'll make a specific class for this.

File: src/Core/Filter/FrontEndObject/CustomerFilter.php (hunk 5 above)
Comment (mickaelandrieu) [reply src/Core/Filter/FrontEndObject/CustomerFilter.php:59]: well, I prefer we include the hook dispatcher as dependency and do:

```php
$whitelist = $this->hookDispatcher->dispatch('actionXxx', ['whitelist' => ... ]);
```

no more global static used, and easier to unit test :+1:

Comment (kpodemski) [reply src/Core/Filter/FrontEndObject/CustomerFilter.php:59]: @mickaelandrieu yeah sure, i forgot about `hookDispatcher`, you get the idea tho :)

**Comment by kpodemski** [comment]
`$this->get('product_front_end_filter')->whitelist(array('myCustomProperty'));`

I love that idea. Overall, i'm all in for having more and more PrestaShop layers as a service, this would be great and add so many possibilities...

**Comment by LittleBigDev** [comment]
Reviewed 8 of 14 files at r1, 1 of 1 files at r2, 8 of 8 files at r3.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[src/Core/Filter/HashMapWhitelistFilter.php, line 140 at r3](https://reviewable.io:443/reviews/prestashop/prestashop/8803#-L6pjqsNzkoaJETQvrRA:-L6pjqsOiKZ4ERpWPXds:bd1ogvh) ([raw file](https://github.com/prestashop/prestashop/blob/6d68d3899d869ed37dff8464557bda2537ccac43/src/Core/Filter/HashMapWhitelistFilter.php#L140)):*
> ```PHP
> 
>     /**
>      * Returns the ested filters, indexed by $keyToKeep
> ```

typo in "nested"

---

*[src/Core/Filter/FrontEndObject/CustomerFilter.php, line 59 at r1](https://reviewable.io:443/reviews/prestashop/prestashop/8803#-L62We25B5hHf5kjT641-r1-59:-L6pluuVAaAgf7TpJa07:b-4l5e5g) ([raw file](https://github.com/prestashop/prestashop/blob/79d9eb9854966a6ca6a68b11be472a2ef2376d15/src/Core/Filter/FrontEndObject/CustomerFilter.php#L59)):*
<details><summary><i>Previously, kpodemski (Krystian Podemski) wrote…</i></summary><blockquote>

@mickaelandrieu yeah sure, i forgot about `hookDispatcher`, you get the idea tho :)
</blockquote></details>

I agree with this suggestion

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8803)*

**Comment by LittleBigDev** [comment]
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[src/Core/Filter/FrontEndObject/CustomerFilter.php, line 41 at r3](https://reviewable.io:443/reviews/prestashop/prestashop/8803#-L6pqJk7M7Heso45M3Dx:-L6pqJk7M7Heso45M3Dy:b9i0ca8) ([raw file](https://github.com/prestashop/prestashop/blob/6d68d3899d869ed37dff8464557bda2537ccac43/src/Core/Filter/FrontEndObject/CustomerFilter.php#L41)):*
> ```PHP
>         $whitelist = array(
>             "addresses",
>             "ape",
> ```

Did you mean "age" ?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8803)*

**Comment by kpodemski** [comment]
@LittleBigDev SIRET/APE, not age ;)

**Comment by LittleBigDev** [comment]
Oh 😅

**Comment by eternoendless** [comment]
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[src/Core/Filter/HashMapWhitelistFilter.php, line 140 at r3](https://reviewable.io:443/reviews/prestashop/prestashop/8803#-L6pjqsNzkoaJETQvrRA:-L6q9QK3AikABr1vecjr:b4c9lyr) ([raw file](https://github.com/prestashop/prestashop/blob/6d68d3899d869ed37dff8464557bda2537ccac43/src/Core/Filter/HashMapWhitelistFilter.php#L140)):*
<details><summary><i>Previously, LittleBigDev (Bastien Bieri) wrote…</i></summary><blockquote>

typo in "nested"
</blockquote></details>

:thumbsup: 

---

*[src/Core/Filter/FrontEndObject/CustomerFilter.php, line 59 at r1](https://reviewable.io:443/reviews/prestashop/prestashop/8803#-L62We25B5hHf5kjT641-r1-59:-L6q9bC5gwn6VsFgIYXD:bqt91v3) ([raw file](https://github.com/prestashop/prestashop/blob/79d9eb9854966a6ca6a68b11be472a2ef2376d15/src/Core/Filter/FrontEndObject/CustomerFilter.php#L59)):*
<details><summary><i>Previously, LittleBigDev (Bastien Bieri) wrote…</i></summary><blockquote>

I agree with this suggestion
</blockquote></details>

There's no need for a specific hook as you can get the service instance and add/remove items to/from the whitelist. You only need a hook that's launched at the right time (before `FrontControllerCore::buildFrontEndObject()`). I can't find a proper hook for that though, so I'm going to add it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8803)*

**Comment by LittleBigDev** [comment]
:lgtm:

---

Reviewed 3 of 3 files at r4.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8803#-:-L6qP5fCgiHzlTpcq2Ko:bnfp4nl)*

**Review by LittleBigDev — DISMISSED** [review]
Thank you @eternoendless 

Maybe a second review should be done to double check the hooks part

**Comment by mickaelandrieu** [comment]
How about the performance ? Adding a hook in frontend is always time consuming

**Comment by kpodemski** [comment]
@mickaelandrieu well, these are three hooks on not dynamic elements, right? i don't know if we loose that much performance on this, tbh this is yet another feature where we would need something like "cached" hook, because there's literally no reason to fire it every time

**Comment by mickaelandrieu** [comment]
@kpodemski yup but if an already existing hook is enough, I dont see the real value of adding a specific hook.

**Comment by eternoendless** [comment]
Tested it with Blackfire:
https://blackfire.io/profiles/compare/2a9d82c2-286d-432e-aafc-cca06bb05909/graph

There's no difference in performance with or without the hooks, at least as long as there's no module hooked to them. Differences in I/O and CPU can only be explained by the fact that I have a thousand things running on my Mac.

**Comment by eternoendless** [comment]
And that's two hooks, not three 🙂

**Review by mickaelandrieu — DISMISSED** [review]

**Comment by marionf** [comment]
There is always some data displayed in the source code

**Comment by LittleBigDev** [comment]
:lgtm:

---

Reviewed 8 of 8 files at r5.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8803#-:-L7Z7DF4DYQPrnNPWIZA:bnfp4nl)*

**Review by LittleBigDev — APPROVED** [review]
:+1:

**Comment by eternoendless** [comment]
@mickaelandrieu WDYT?

**Review by mickaelandrieu — APPROVED** [review]

**Comment by mickaelandrieu** [comment]
Thank you @eternoendless ! And thanks everyone for reviews

**Comment by kpodemski** [comment]
docs

remember about the docs

**Comment by mickaelandrieu** [comment]
@kpodemski feel free to help us, you can start from `hugo` branch of docs repository :+1:

**Comment by eternoendless** [comment]
I'm writing an article for build, devdocs will follow

**Comment by kpodemski** [comment]
@mickaelandrieu yeah, i need to get familiar with this Hugo, every year some fancy new stuff for docs ;p

**Comment by mickaelandrieu** [comment]
@kpodemski in the end it's only "hugo serve" to render the docs in local and markdown :)
If you need help, I'll be available on gitter and I'm working on docs today ^^
